antismash/common/secmet/locations.py

```diff
@@ -16,59 +16,59 @@
+
+def convert_protein_position_to_dna(start: int, end: int, location: FeatureLocation, codon_start: int) -> Tuple[int, int]:
     """ Convert a protein position to a nucleotide sequence position for use in generating
         new FeatureLocations from existing FeatureLocations and/or CompoundLocations.
 
         Arguments:
             position: the position in question, must be contained by the location
             location: the location of the related feature, for handling introns/split locations
-
+            codon_start: the offset with which the protein coding sequence starts
         Returns:
             an int representing the calculated DNA location
     """
     if not 0 <= start < end <= len(location) // 3:
         raise ValueError("Protein positions %d and %d must be contained by %s" % (start, end, location))
-    if location.strand == -1:
-        dna_start = location.start + len(location) - end * 3
-        dna_end = location.start + len(location) - start * 3
-    else:
-        dna_start = location.start + start * 3
-        dna_end = location.start + end * 3
+    dna_start = start * 3
+    dna_end = end * 3
 
-    # only CompoundLocations are complicated
-    if not isinstance(location, CompoundLocation):
-        if not location.start <= dna_start < dna_end <= location.end:
-            raise ValueError(("Converted coordinates %d..%d "
-                              "out of bounds for location %s") % (dna_start, dna_end, location))
-        return dna_start, dna_end
+    # typically codon_start = 1, the pythonic way
+    codon_start_offset = codon_start - 1
 
-    parts = sorted(location.parts, key=lambda x: x.start)
-    gap = 0
-    last_end = parts[0].start
+    processed = 0
     start_found = False
     end_found = False
-    for part in parts:
+    for part in location.parts:
         if start_found and end_found:
             break
-        gap += part.start - last_end
-        if not start_found and dna_start + gap in part:
+        if not start_found and dna_start + codon_start_offset < len(part) + processed:
             start_found = True
-            dna_start = dna_start + gap
-        if not end_found and dna_end + gap - 1 in part:
+            if part.strand == -1:
+                dna_start = part.bio_start - dna_start - codon_start_offset + processed
+            else:
+                dna_start = part.bio_start + dna_start + codon_start_offset - processed
+        if not end_found and dna_end <= len(part) + processed:
             end_found = True
-            dna_end = dna_end + gap
-
-        last_end = part.end
+            if part.strand == -1:
+                dna_end = part.bio_start - dna_end - codon_start_offset + processed
+            else:
+                dna_end = part.bio_start + dna_end + codon_start_offset - processed
+        processed += len(part)
 
     assert start_found
     assert end_found
 
-    if not location.start <= dna_start < dna_end <= location.end:
```

Comment:
This is a complex issue and i am still thinking what to do here.

reasons... i can think of at least:
1. Tools that, when extracting a part of the sequence and when this extracted part truncates the CDS to a partial one, do not update neither "/codon_start" nor "/translation" tags.
2. NCBI annotation pipeline, which sometimes annotates a "/translation" which is one amino acid longer than it is supposed to be. Mostly those are truncated genes which have a nucleotide length of 3*N+2, where N is the number of aminoacids. Instead of N, the pipeline annotates "/translation=" to N+1 amino acids, because the trailing 2 bp in many cases can be unambiguously translated to an amino acid, even though we do not know the third base. What can we do...

What about? Checking for this when examining translations? Verifying if a translation fits a CDS feature can be done there. Then this check can happily live here.

Reply:
Ensuring that the translation matches the rest of the annotations seems like a good plan.

In general, the aim is that once an input record is fully parsed and a secmet Record exists, it should be in a valid state. If the input contains issues that we can't solve simply (like a missing translation), then we stop then and alert the user that their input is in a bad state and let them decide how it should be fixed, since they have much more information about their input than we do.